src/registry.rs

```diff
@@ -0,0 +1,672 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! ObjectStoreRegistry holds object stores at runtime with a URL for each store.
+//! The registry serves as a cache for object stores to avoid repeated creation.
+//! It also lets you convert an [`ObjectStore`] back to its registered URL via
+//! `get_prefix`.
+use crate::{parse_url, ObjectStore};
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+use url::Url;
+
+/// [`ObjectStoreRegistry`] maps a URL prefix to an [`ObjectStore`] instance. The definition of
+/// a URL prefix depends on the [`ObjectStoreRegistry`] implementation. See implementation docs
+/// for more details.
+pub trait ObjectStoreRegistry: Send + Sync + std::fmt::Debug + 'static {
+    /// Register a new store for any URL that begins with `prefix`
+    ///
+    /// If a store with the same prefix existed before, it is replaced and returned
+    fn register_store(
+        &self,
+        prefix: &Url,
+        store: Arc<dyn ObjectStore>,
+    ) -> Option<Arc<dyn ObjectStore>>;
+
+    /// Get a store for the provided URL. The definition of a "suitable store" depends on
+    /// the [`ObjectStoreRegistry`] implementation. See implementation docs for more details.
+    ///
+    /// If no [`ObjectStore`] is found for the `url`, an [`ObjectStore`] may be lazily be
+    /// created and registered. The logic for doing so is left to each [`ObjectStoreRegistry`]
+    /// implementation.
+    fn get_store(&self, url: &Url) -> Option<Arc<dyn ObjectStore>>;
+
+    /// Given one of the `Arc<dyn ObjectStore>`s you registered, return its URL.
+    fn get_prefix(&self, store: Arc<dyn ObjectStore>) -> Option<Url>;
+
+    /// List all registered store prefixes
+    fn get_store_prefixes(&self) -> Vec<Url>;
```

Comment:
Similar to the above, I'm not sure about this method. It isn't well defined what this means, as it is up to the implementation how it wishes to key stores, which makes this abstraction leaky